core/src/token_info.rs

```diff
@@ -1,23 +1,74 @@
 use anyhow::Result;
-use futures::future::BoxFuture;
+use futures::future::{BoxFuture, FutureExt as _};
 use lazy_static::lazy_static;
-#[cfg(test)]
-use mockall::automock;
 use std::{collections::HashMap, num::NonZeroU128};
 
 use crate::models::TokenId;
 pub mod cached;
 pub mod hardcoded;
 pub mod onchain;
 
-#[cfg_attr(test, automock)]
 pub trait TokenInfoFetching: Send + Sync {
     /// Retrieves some token information from a token ID.
     fn get_token_info<'a>(&'a self, id: TokenId) -> BoxFuture<'a, Result<TokenBaseInfo>>;
 
+    /// Retrieves all token information.
+    /// Default implementation calls get_token_info for each token and ignores errors.
+    fn get_token_infos<'a>(
+        &'a self,
+        ids: &'a [TokenId],
+    ) -> BoxFuture<'a, Result<HashMap<TokenId, TokenBaseInfo>>> {
+        async move {
+            let mut result = HashMap::new();
+            for id in ids {
+                match self.get_token_info(*id).await {
+                    Ok(info) => {
+                        result.insert(*id, info);
+                    }
+                    Err(err) => log::warn!("failed to get token info for {}: {:?}", id, err),
+                }
+            }
+            Ok(result)
+        }
+        .boxed()
+    }
+
     /// Returns a vector with all the token IDs available
     fn all_ids<'a>(&'a self) -> BoxFuture<'a, Result<Vec<TokenId>>>;
 }
+
+// mockall workaround https://github.com/asomers/mockall/issues/134
+#[cfg(test)]
+mod mock {
```

Comment:
I know we had this discussion before, but just re-checking that we feel more comfortable maintaining this workaround than making the trait take TokenId (which is an u16) by value.

Reply:
It's `&[TokenId]` vs `Vec<TokenId>`.